### PR TITLE
[stdlib] Random unification

### PR DESCRIPTION
**Pull request by Azoy**
Alright, lets try this again!
This is the implementation for [SE-0202](https://github.com/apple/swift-evolution/blob/master/proposals/0202-random-unification.md)

### CODE REVIEW

**Comment by Azoy** [comment]
cc: @airspeedswift

File: stdlib/public/stubs/LibcShims.cpp
Comment (natecook1000) [inline stdlib/public/stubs/LibcShims.cpp]: Is this going to break the build for Win32 platforms?

**Comment by natecook1000** [comment]
@swift-ci Please smoke test

**Comment by natecook1000** [comment]
@swift-ci Please smoke benchmark

**Review by natecook1000 — CHANGES_REQUESTED** [review]
I just have this one Q, then LGTM pending tests. I have some more documentation edits ready to go, but let's get this merged first!

File: stdlib/public/stubs/LibcShims.cpp
Comment (Azoy) [reply stdlib/public/stubs/LibcShims.cpp]: Yes and no, this does break builds if they're compiling the stdlib. I can probably substitute this in with a warning directive instead.

File: stdlib/public/core/Random.swift
Comment (milseman) [inline stdlib/public/core/Random.swift]: I wonder if it would make sense to split off the rest of this function into a non-`@inlinable` `@inline(never)` internal method, and `_fastPath` the branch above. This could make it more likely that the 99.99% use case and fast-path is more reliably inlined and/or specialized. CC @eeckstein for thoughts.

Also, how is the rest of this code tested?

File: stdlib/public/core/Random.swift
Comment (milseman) [inline stdlib/public/core/Random.swift]: Could you add a comment explaining the implementation strategy and the reason you chose it?

File: stdlib/public/stubs/LibcShims.cpp
Comment (milseman) [reply stdlib/public/stubs/LibcShims.cpp]: CC @compnerd

File: stdlib/public/stubs/LibcShims.cpp
Comment (milseman) [inline stdlib/public/stubs/LibcShims.cpp]: I'm not sure if this code should apply to `arc4random` on Darwin platforms, which claims it never returns an error code.

**Review by milseman — APPROVED** [review]
Overall LGTM, just some minor comments.

I'll let @moiseev review the floating point and integer code.

File: stdlib/public/stubs/LibcShims.cpp
Comment (Azoy) [reply stdlib/public/stubs/LibcShims.cpp]: This does not apply to `arc4random` on Darwin platforms and won't be executed during runtime. I can separate Darwin's solution outside of this one if that clears any confusion.

File: stdlib/public/stubs/LibcShims.cpp
Comment (CodaFi) [inline stdlib/public/stubs/LibcShims.cpp]: Nit: Please re-outdent the conditional preprocessor directives.

File: stdlib/public/stubs/LibcShims.cpp
Comment (CodaFi) [inline stdlib/public/stubs/LibcShims.cpp]: Is a blocking read from the OS's entropy source the right behavior here?  For applications like server side Swift we can pretty well expect a fast response from the OS because it'll have initialized its entropy sources by the time a Swift process is up.  But for embedded platforms this call might take as long as minutes if the OS deems there's insufficient entropy.

There's pros and cons either way, I would just like to see a rationale for this somewhere for reference.

Comment (Azoy) [reply stdlib/public/stubs/LibcShims.cpp]: The version of `getrandom` we're using here only blocks on OS startup where `/dev/urandom` may not have been initialized yet. After `/dev/urandom` is initialized, `getrandom` will always call into it, which makes this call non-blocking 99.99% of the time. The [proposal](https://github.com/apple/swift-evolution/blob/master/proposals/0202-random-unification.md#shouldnt-this-fallback-on-something-more-secure-at-times-of-low-entropy) explains this and shares that Python's implementation uses `getrandom()` as well, but falls back on `/dev/urandom` if `getrandom()` deems it needed to block on OS startup. We could do something similar.

Comment (CodaFi) [reply stdlib/public/stubs/LibcShims.cpp]: > The proposal explains this and shares that Python's implementation uses getrandom() as well, but falls back on /dev/urandom if getrandom() deems it needed to block on OS startup. We could do something similar.

We could, but doesn't Python 3.6 do what you've got here now?

Comment (Azoy) [reply stdlib/public/stubs/LibcShims.cpp]: Ah you're right, 3.6 does what we have here and 3.5 fell back on reading `/dev/urandom`.

File: stdlib/public/core/Random.swift
Comment (natecook1000) [reply stdlib/public/core/Random.swift]: We have the `DoubleWidth` type here, we could test a 128-bit integer type with that: https://github.com/apple/swift/blob/master/test/Prototypes/DoubleWidth.swift.gyb

If we want to test anything that isn't a power of two wide, that'll be an additional thing to develop.

**Review by natecook1000 — APPROVED** [review]

File: stdlib/public/stubs/LibcShims.cpp
Comment (CodaFi) [inline stdlib/public/stubs/LibcShims.cpp]: `getentropy` is usually used for seeding.  [zx_cprng_draw](https://fuchsia.googlesource.com/zircon/+/HEAD/docs/syscalls/cprng_draw.md) seems like the function we oughta call here but I'm no Googler so @zbowling?

**Comment by zbowling** [comment]
Yeah, zx_cprng_draw is the lowest level way to get RNG on Fuchsia.
getentropy should also work as it calls zx_cprng_draw now and now libc++
std::random_device should have support for getentropy as a backend (including
our version in Fuchsia) from a patch we landed up stream.  See:
https://reviews.llvm.org/D40319

zx_cprng_draw gives you more control when the entropy pool is empty than
the genentropy API does (very unlikely) but we generally prefer it in code
repos outside of our own use any standard APIs when possible today as our
APIs are subject to change (although getting more and more unlikely over
time).


Zac Bowling

**Comment by airspeedswift** [comment]
Unless there are any last-minute concerns, I'm going to merge this as it stands, and items from this PR can be resolved by follow-up commits.

**Comment by airspeedswift** [comment]
@swift-ci please test and merge

**Comment by natecook1000** [comment]
@swift-ci Please test and merge

**Comment by airspeedswift** [comment]
@swift-ci please test macOS platform

**Comment by airspeedswift** [comment]
Hmm. Seems to be failing reproducibly on the simulator, but only specifically on the Double testing.

**Comment by Azoy** [comment]
I can look into it when I get home, although that won’t be for another 8 hours 😕

File: stdlib/public/core/Random.swift
Comment (lorentey) [inline stdlib/public/core/Random.swift]: I wonder how much resilience costs us for `Random`; `Random.default.next()` involves an indirectly-sized temporary and three resilient function calls, when we only really need a single call.

Do we expect that `Random` may become non-empty in the future? If not, we should make it `@_fixed_layout`, and we should make the initializer and `.default`'s getter and setter inlinable.

File: stdlib/public/core/Random.swift
Comment (lorentey) [inline stdlib/public/core/Random.swift]: Alas, this isn't right; `FixedWidthInteger` types aren't necessarily bitwise initializable like this. 

It may very well be that we can't easily do better than to provide four additional concrete overloads for `UInt8`, `UInt16`, `UInt32` and `UInt`, and leave the generic case to the default implementation. 😞


A more complicated way to resolve this would be to implement these `next()` overloads using a new (defaulted) `_random(using:)` requirement on `FixedWidthInteger`, which the standard fixed width integers could implement using `_stdlib_random`. To make this work, we should expose a customizable `_fillBuffer` operation on `RandomNumberGenerator`.

```swift
protocol RandomNumberGenerator {
  mutating func next() -> UInt64
  // FIXME: De-underscore after swift-evolution amendment
  mutating func _fillBuffer(_ buffer: UnsafeMutableRawBufferPointer)
} 
extension RandomNumberGenerator {
  public mutating func _fillBuffer(_ buffer: UnsafeMutableRawBufferPointer) {
    // Implementation in terms of next()
  }
  // FIXME: Remove after swift-evolution amendment
  @inlinable
  public mutating next<T: FixedWidthInteger & UnsignedInteger>() -> T {
    return T._random(using: &self)
  }
}
extension Random {
   @effects(releasenone)
   public mutating func _fillBuffer(_ buffer: UnsafeMutableRawBufferPointer) {
    _stdlib_random(buffer)
  }
}

protocol FixedWidthInteger: ... {
  ...
  // FIXME: De-underscore after swift-evolution amendment
  static func _random<R: RandomNumberGenerator>(using generator: inout R) -> Self
}
extension FixedWidthInteger {
  @inlinable
  public static func _random<R: RandomNumberGenerator>(using generator: inout R) -> Self {
    // Generic implementation with &<< and | (or +)
  }
}

extension UInt8 {
  @inlinable
  public static func _random<R: RandomNumberGenerator>(using generator: inout R) -> UInt8 {
    var result: UInt8 = 0
    withUnsafeMutableBytes(of: &result) { generator._fillBuffer($0) }
    return result
  }
}
// etc. for the other standard fixed width integer types
```

File: stdlib/public/core/Random.swift
Comment (lorentey) [inline stdlib/public/core/Random.swift]: If filling a buffer with random bytes is the primitive operation around which we're building the default RNG, it should probably be exposed as a customizable public operation on `RandomNumberGenerator`. (I.e., I think it should be a requirement, with a default implementation in terms of the 64-bit `next`.)

(Update: I see you've already made the case for this on swift-evolution. 👍)

**Review by lorentey — CHANGES_REQUESTED** [review]
Most of my comments aren't important enough to block this PR, except for the issue with the generic implementation of `Random.next()`. I'm requesting changes for that one issue only; although I'm also fine with landing this now and fixing that in a followup -- but it should be fixed soon.

File: stdlib/public/core/Random.swift
Comment (airspeedswift) [reply stdlib/public/core/Random.swift]: Yeah I see that as a good future enhancement.

File: stdlib/public/core/Random.swift
Comment (airspeedswift) [reply stdlib/public/core/Random.swift]: Ah good spot – the decision for the Apple implementation is to use one consistent implementation regardless of platform version. The caveats of OS version determining the method of generation exist today for `arc4random` and should extend to this implementation.

File: stdlib/public/core/Random.swift
Comment (Azoy) [reply stdlib/public/core/Random.swift]: Is there a way to mark a private initializer as `@inlinable` or something similar? I can't use a private initializer in an `@inlinable` getter/setter for a var, but I can if I make the initializer internal with `@usableFromInline`. I wonder if we have to make this change in order to achieve `@inlinable` for the static var.

Comment (lorentey) [reply stdlib/public/core/Random.swift]: If we're okay with making it inlinable, I think `@inlinable internal` is the way to do it.

File: stdlib/public/core/Integers.swift.gyb
Comment (stephentyrone) [inline stdlib/public/core/Integers.swift.gyb]: It might be worth opening jira / radar to track making this less spectacularly inefficient for bignums at some future point?

**Comment by lorentey** [comment]
@swift-ci please smoke test

**Comment by lorentey** [comment]
@swift-ci please smoke test

**Review by lorentey — APPROVED** [review]

**Comment by lorentey** [comment]
@Azoy By a (happy?) accident, I pushed two commits directly on your branch instead of my own fork; they fix the FixedWidthInteger issue I noticed.

Provided the bots are in a cooperating mood, I want to re-do the tests to try reproducing the failure on the 32-bit Simulator -- I can't reproduce it myself with the latest changes.

**Comment by lorentey** [comment]
@swift-ci smoke test macOS platform

**Comment by Azoy** [comment]
@lorentey I actually fixed those in my commit last night!

**Comment by lorentey** [comment]
@Azoy Nice!

Unless anyone spots an issue with the last round of changes, LGTM! We just need to wait for the bots to come back.

**Comment by stephentyrone** [comment]
@lorentey Agree. Assuming tests pass, let's land this so people can start using it. We'll find the problems faster that way anyway =)

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
(needs full test to get the simulators)

**Review by xwu — COMMENTED** [review]
Sorry to be late here. Some initial comments.

File: stdlib/public/core/Bool.swift
Comment (stephentyrone) [reply stdlib/public/core/Bool.swift]: This isn't totally insane, because for very weak RNGs middle bits have better randomness properties than low bits, but really no one should be using those ...

Comment (Azoy) [reply stdlib/public/core/Bool.swift]: Nate sent this in the initial pr:
> For some random generators the low bits have higher levels of repetition than the higher bits do. Could we change this to (generator.next() >> 17) & 1 == 0?

Comment (xwu) [reply stdlib/public/core/Bool.swift]: Got it; seemed strange because `17` is pretty arbitrary. If we want to use high bits, `Int(bitPattern: generator.next()) < 0` would be less mystifying, IMO.

Comment (stephentyrone) [reply stdlib/public/core/Bool.swift]: @xwu you don't want the *high* bits either, those can also be weak. You really want some "middle" bits. This is worth a comment, though.

File: stdlib/public/core/Random.swift
Comment (xwu) [reply stdlib/public/core/Random.swift]: Ah yes, I didn't study the preceding lines in great detail. Agree that there's room for performance improvement.

**Comment by natecook1000** [comment]
@swift-ci Please test and merge

File: stdlib/public/core/Bool.swift
Comment (natecook1000) [reply stdlib/public/core/Bool.swift]: Go ahead, try to name a number more random than `17`.

Comment (natecook1000) [reply stdlib/public/core/Bool.swift]: FWIW, according to some cursory testing, the `GKRandomSource` types use `31` for the offset in `nextBool()` (i.e., the high bit, like @xwu suggested).

**Comment by airspeedswift** [comment]
@swift-ci please test macOS platform

**Comment by lorentey** [comment]
This looks ready to merge 🎉

File: stdlib/public/core/Integers.swift.gyb
Comment (lorentey) [reply stdlib/public/core/Integers.swift.gyb]: I think adding required `BinaryInteger` initializers to initialize integers from a given word or sequence of words would resolve this. I filed https://bugs.swift.org/browse/SR-7648; we should write up a swift-evolution pitch/proposal at some point. (cc @moiseev)

(The code above is defined on `FixedWidthInteger`, which uses a similar shifting loop for its own generic integer conversions. Frustratingly, `FixedWidthInteger` is currently impossible to implement using only public API: it requires an underscored version of the single-word initializer, without providing a default implementation.)

Comment (stephentyrone) [reply stdlib/public/core/Integers.swift.gyb]: Thanks for writing that up!

**Pull request merged**
